File: shops-app/src/services/analytics-tracking.ts
/**
 * Analytics Tracking Service
 * Tracks user interactions, clicks, page views, and more
 */

import api from './api';
import { useAuthStore } from '@/store/useAuth';

interface ClickEventData {
  element_id?: string;
  element_class?: string;
  element_type: string;
  text?: string;
  x: number;
  y: number;
  page_url: string;
  page_width: number;
  page_height: number;
}

interface ActivityEventData {
  action_type: string;
  action_category: string;
  resource_id?: string;
  page_url: string;
  search_query?: string;
}

class AnalyticsTracker {
  private enabled = true;
  private sessionToken: string = '';
  private pageViewTimeout: NodeJS.Timeout | null = null;
  private pageStartTime: number = 0;
  private maxScrollDepth: number = 0;
  private scrollCheckInterval: NodeJS.Timeout | null = null;

  constructor() {
    this.initializeSession();
    this.setupClickTracking();
    this.setupPageTracking();
    this.setupScrollTracking();
    this.setupTimeTracking();
  }

  /**
   * Initialize session tracking
   */
  private initializeSession() {
    // Generate or retrieve session token
    const storedToken = localStorage.getItem('analytics_session_token');
    if (storedToken) {
      this.sessionToken = storedToken;
    } else {
      this.sessionToken = this.generateSessionToken();
      localStorage.setItem('analytics_session_token', this.sessionToken);
    }
  }

  /**
   * Generate unique session token
   */
  private generateSessionToken(): string {
    return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Setup global click event tracking
   */
  private setupClickTracking() {
    if (!this.enabled) return;

    document.addEventListener('click', (e) => {
      const target = e.target as HTMLElement;
      if (!target) return;

      // Get element info
      const elementId = target.id || '';
      const elementClass = target.className || '';
      const elementType = target.tagName.toLowerCase();
      const text = target.textContent?.substring(0, 50) || '';

      // Get click position
      const x = e.clientX;
      const y = e.clientY;
      const pageUrl = window.location.pathname;
      const pageWidth = window.innerWidth;
      const pageHeight = window.innerHeight;

      // Track click
      this.trackClick({
        element_id: elementId,
        element_class: elementClass,
        element_type: elementType,
        text: text,
        x,
        y,
        page_url: pageUrl,
        page_width: pageWidth,
        page_height: pageHeight,
      });
    });
  }

  /**
   * Setup scroll depth tracking
   */
  private setupScrollTracking() {
    if (!this.enabled) return;

    window.addEventListener('scroll', () => {
      const scrollHeight = document.documentElement.scrollHeight - window.innerHeight;
      if (scrollHeight > 0) {
        const scrollPercentage = (window.scrollY / scrollHeight) * 100;
        this.maxScrollDepth = Math.max(this.maxScrollDepth, scrollPercentage);
      }
    });
  }

  /**
   * Setup time on page tracking
   */
  private setupTimeTracking() {
    if (!this.enabled) return;

    this.pageStartTime = Date.now();

    // Track time on page every 30 seconds
    setInterval(() => {
      const timeOnPage = (Date.now() - this.pageStartTime) / 1000; // seconds
      if (timeOnPage > 30) {
        this.trackMetric('time_on_page', {
          seconds: Math.floor(timeOnPage),
          scroll_depth: this.maxScrollDepth,
        });
      }
    }, 30000);

    // Track on page leave
    window.addEventListener('beforeunload', () => {
      const timeOnPage = (Date.now() - this.pageStartTime) / 1000;
      this.trackMetric('page_exit', {
        seconds: Math.floor(timeOnPage),
        scroll_depth: this.maxScrollDepth,
      });
    });
  }

  /**
   * Track page view
   */
  private setupPageTracking() {
    if (!this.enabled) return;

    // Track initial page view
    this.trackPageView();

    // Reset metrics on page change
    const resetMetrics = () => {
      this.maxScrollDepth = 0;
      this.pageStartTime = Date.now();
    };

    // Track page changes (for SPA)
    const originalPushState = window.history.pushState;
    const originalReplaceState = window.history.replaceState;

    window.history.pushState = function (...args) {
      originalPushState.apply(this, args);
      resetMetrics();
      setTimeout(() => this.trackPageView(), 100);
    };

    window.history.replaceState = function (...args) {
      originalReplaceState.apply(this, args);
      resetMetrics();
      setTimeout(() => this.trackPageView(), 100);
    };
  }

  /**
   * Track a click event
   */
  async trackClick(event: ClickEventData) {
    if (!this.enabled) return;

    try {
      // Batch clicks - send in background without blocking
      navigator.sendBeacon?.(
        '/api/v1/analytics/track/click',
        JSON.stringify({
          ...event,
          session_token: this.sessionToken,
          user_id: this.getCurrentUserId(),
          timestamp: new Date().toISOString(),
        })
      );
    } catch (err) {
      console.error('Failed to track click:', err);
    }
  }

  /**
   * Track page view
   */
  async trackPageView() {
    if (!this.enabled) return;

    // Debounce page view tracking
    if (this.pageViewTimeout) clearTimeout(this.pageViewTimeout);

    this.pageViewTimeout = setTimeout(() => {
      try {
        const userId = this.getCurrentUserId();
        if (!userId) return; // Only track authenticated users

        api.post('/analytics/track/activity', {
          action_type: 'view_page',
          action_category: 'navigation',
          page_url: window.location.pathname,
          session_token: this.sessionToken,
        }).catch(() => {}); // Silently fail - don't interrupt user experience
      } catch (err) {
        console.error('Failed to track page view:', err);
      }
    }, 1000); // Debounce by 1s
  }

  /**
   * Track search action
   */
  async trackSearch(query: string) {
    if (!this.enabled) return;

    try {
      const userId = this.getCurrentUserId();
      if (!userId) return;

      await api.post('/analytics/track/activity', {
        action_type: 'search',
        action_category: 'search',
        search_query: query,
        page_url: window.location.pathname,
        session_token: this.sessionToken,
      });
    } catch (err) {
      console.error('Failed to track search:', err);
    }
  }

  /**
   * Track listing view
   */
  async trackListingView(listingId: string | number) {
    if (!this.enabled) return;

    try {
      const userId = this.getCurrentUserId();
      if (!userId) return;

      await api.post('/analytics/track/activity', {
        action_type: 'view_listing',
        action_category: 'listing',
        resource_id: listingId.toString(),
        page_url: window.location.pathname,
        session_token: this.sessionToken,
      });
    } catch (err) {
      console.error('Failed to track listing view:', err);
    }
  }

  /**
   * Track add to cart
   */
  async trackAddToCart(listingId: string | number) {
    if (!this.enabled) return;

    try {
      const userId = this.getCurrentUserId();
      if (!userId) return;

      await api.post('/analytics/track/activity', {
        action_type: 'add_to_cart',
        action_category: 'cart',
        resource_id: listingId.toString(),
        page_url: window.location.pathname,
        session_token: this.sessionToken,
      });
    } catch (err) {
      console.error('Failed to track add to cart:', err);
    }
  }

  /**
   * Track purchase
   */
  async trackPurchase(orderId: string | number) {
    if (!this.enabled) return;

    try {
      const userId = this.getCurrentUserId();
      if (!userId) return;

      await api.post('/analytics/track/activity', {
        action_type: 'purchase',
        action_category: 'order',
        resource_id: orderId.toString(),
        page_url: window.location.pathname,
        session_token: this.sessionToken,
      });
    } catch (err) {
      console.error('Failed to track purchase:', err);
    }
  }

  /**
   * Get current authenticated user ID
   */
  private getCurrentUserId(): number | null {
    try {
      const user = useAuthStore.getState().user;
      return user?.id || null;
    } catch {
      return null;
    }
  }

  /**
   * Track custom metrics
   */
  async trackMetric(metricName: string, data: any) {
    if (!this.enabled) return;

    try {
      const userId = this.getCurrentUserId();
      if (!userId) return;

      await api.post('/analytics/track/metric', {
        metric_name: metricName,
        metric_data: JSON.stringify(data),
        page_url: window.location.pathname,
        session_token: this.sessionToken,
      }).catch(() => {});
    } catch (err) {
      console.error('Failed to track metric:', err);
    }
  }

  /**
   * Enable/disable tracking
   */
  setEnabled(enabled: boolean) {
    this.enabled = enabled;
  }
}

// Export singleton instance
export const analyticsTracker = new AnalyticsTracker();

// Export for use in components
export default {
  trackSearch: (query: string) => analyticsTracker.trackSearch(query),
  trackListingView: (id: string | number) => analyticsTracker.trackListingView(id),
  trackAddToCart: (id: string | number) => analyticsTracker.trackAddToCart(id),
  trackPurchase: (id: string | number) => analyticsTracker.trackPurchase(id),
  setEnabled: (enabled: boolean) => analyticsTracker.setEnabled(enabled),
};

File: backend/app/api/api_v1/endpoints/analytics_tracking.py
"""Analytics event tracking endpoints (lightweight, async)."""

from fastapi import APIRouter, Depends, Request, BackgroundTasks
from sqlmodel import Session
from datetime import datetime
from typing import Optional
from app.api import deps
from app.models.analytics import UserActivity, ClickEvent, UserSession
from app.models.user import User
import json

router = APIRouter()


@router.post("/track/click")
async def track_click(
    request: Request,
    background_tasks: BackgroundTasks,
    db: Session = Depends(deps.get_db),
):
    """
    Track click events (lightweight, batched in background).
    Can be called without authentication (anonymous tracking).
    """
    try:
        data = await request.json()

        # Process in background to avoid blocking
        async def save_click():
            try:
                click = ClickEvent(
                    user_id=data.get("user_id"),
                    session_id=data.get("session_token"),
                    element_id=data.get("element_id"),
                    element_class=data.get("element_class"),
                    element_type=data.get("element_type", "unknown"),
                    text=data.get("text"),
                    x=int(data.get("x", 0)),
                    y=int(data.get("y", 0)),
                    page_url=data.get("page_url", ""),
                    page_width=int(data.get("page_width", 1024)),
                    page_height=int(data.get("page_height", 768)),
                    timestamp=datetime.utcnow(),
                )
                db.add(click)
                db.commit()
            except Exception as e:
                print(f"Error saving click event: {e}")

        background_tasks.add_task(save_click)
        return {"status": "received"}

    except Exception as e:
        print(f"Error in track_click: {e}")
        return {"status": "error"}


@router.post("/track/metric")
async def track_metric(
    request: Request,
    background_tasks: BackgroundTasks,
    current_user: Optional[User] = Depends(deps.get_current_active_user),
    db: Session = Depends(deps.get_db),
):
    """
    Track custom metrics (time on page, scroll depth, etc).
    """
    try:
        data = await request.json()

        # Process in background
        async def save_metric():
            try:
                activity = UserActivity(
                    user_id=current_user.id if current_user else None,
                    session_id=data.get("session_token"),
                    action_type=data.get("metric_name", "metric"),
                    action_category="metrics",
                    page_url=data.get("page_url", ""),
                    event_metadata=data.get("metric_data", "{}"),
                    timestamp=datetime.utcnow(),
                    ip_address=request.client.host if request.client else None,
                )
                db.add(activity)
                db.commit()
            except Exception as e:
                print(f"Error saving metric: {e}")

        background_tasks.add_task(save_metric)
        return {"status": "received"}

    except Exception as e:
        print(f"Error in track_metric: {e}")
        return {"status": "error"}


@router.post("/track/activity")
async def track_activity(
    request: Request,
    background_tasks: BackgroundTasks,
    current_user: Optional[User] = Depends(deps.get_current_active_user),
    db: Session = Depends(deps.get_db),
):
    """
    Track user activities (search, view, purchase, etc).
    Works for both authenticated and anonymous users.
    """
    try:
        data = await request.json()

        # Process in background
        async def save_activity():
            try:
                activity = UserActivity(
                    user_id=current_user.id if current_user else None,
                    session_id=data.get("session_token"),
                    action_type=data.get("action_type", "unknown"),
                    action_category=data.get("action_category", "other"),
                    resource_id=data.get("resource_id"),
                    page_url=data.get("page_url", ""),
                    search_query=data.get("search_query"),
                    timestamp=datetime.utcnow(),
                    ip_address=request.client.host if request.client else None,
                )
                db.add(activity)
                db.commit()

                # Update conversion funnel if user is authenticated
                if current_user:
                    from app.models.analytics import ConversionFunnel

                    funnel = db.query(ConversionFunnel).filter(
                        ConversionFunnel.user_id == current_user.id
                    ).first()

                    if not funnel:
                        funnel = ConversionFunnel(user_id=current_user.id)
                        db.add(funnel)

                    # Update funnel stage based on action
                    if activity.action_type == "search":
                        if not funnel.first_search_at:
                            funnel.first_search_at = datetime.utcnow()
                            funnel.current_stage = "search"
                    elif activity.action_type == "view_listing":
                        if not funnel.first_view_listing_at:
                            funnel.first_view_listing_at = datetime.utcnow()
                            funnel.current_stage = "view"
                    elif activity.action_type == "add_to_cart":
                        if not funnel.first_add_to_cart_at:
                            funnel.first_add_to_cart_at = datetime.utcnow()
                            funnel.current_stage = "cart"
                    elif activity.action_type == "purchase":
                        if not funnel.first_purchase_at:
                            funnel.first_purchase_at = datetime.utcnow()
                            funnel.current_stage = "purchase"
                            funnel.completed = True
                        elif funnel.repeat_purchase_at is None:
                            funnel.repeat_purchase_at = datetime.utcnow()

                    funnel.updated_at = datetime.utcnow()
                    db.add(funnel)
                    db.commit()

            except Exception as e:
                print(f"Error saving activity: {e}")

        background_tasks.add_task(save_activity)
        return {"status": "received"}

    except Exception as e:
        print(f"Error in track_activity: {e}")
        return {"status": "error"}
